File: 3º ano/Programação Concorrente/Testes e Exames/solução e enunciado teste 2019 (22 maio)/jogo.java
import java.util.Random;

interface Jogo {
    Partida participa() throws InterruptedException ;
}

interface Partida {
    String adivinha(int n);
}

class JogoImpl implements Jogo {
    PartidaImpl p = new PartidaImpl();
    int jogadores = 0;

    public synchronized Partida participa() throws InterruptedException {
        PartidaImpl ps = p;
        jogadores++;
        if (jogadores == 4) {
            notifyAll();
            jogadores = 0;
            ps.start();
            p = new PartidaImpl();
        } else {
            while (p == ps) wait();
        }

        return ps;
    }
}

class PartidaImpl implements Partida {
    int tentativas = 0;
    boolean ganhou = false;
    boolean timeout = false;
    int numero;

    public void start() throw InterruptedException {
        numero = new Random().nextInt(100);
        new Thread(() -> {
            Thread.sleep(60000);
            timeout();
        }).start();
    }

    synchronized void timeout() {
        timeout = true;
    }

    public synchronized String adivinha(int n) {
        if (ganhou) return "PERDEU";
        else if (tentativas >= 100) return "TENTATIVAS";
        else if (timeout) return "TEMPO";
        else {
            tentativas++;
            if (n > numero) return "MENOR";
            else if (n < numero) return "MAIOR";
            else {
                ganhou = true;
                return "GANHOU";
            }
        }

        return "";
    }
}

File: 3º ano/Programação Concorrente/Testes e Exames/solução e enunciado teste 2019 (22 maio)/jogo.erl
-module(jogo).
-export([server/0, jogador/1, jogar/2]).

server() -> spawn(fun() -> jogo([]) end).

jogo(Jogadores) ->
    receive
        {participa, Pid} ->
            NewJogadores = [Pid | Jogadores],
            if
                length(NewJogadores) == 4 ->
                    io:format("Nova Partida: ~p, ~p, ~p, ~p~n", NewJogadores),
                    Partida = spawn(fun() -> initPartida(NewJogadores, false, 0, false, rand:uniform(100)) end), % partida(Jogadores, ganhou, tentativas, tempo)
                    [JogadorPid ! {start, Partida} || JogadorPid <- NewJogadores],
                    jogo([]);
                true ->
                    jogo(NewJogadores)
            end
    end.

timer(Pid) ->
    timer:sleep(20000),
    Pid ! tempo,
    io:fwrite("Tempo acabou.~n").

initPartida(Jogadores, Ganhou, Tentativas, Tempo, Numero) ->
    spawn(fun() -> timer(self()) end),
    partida(Jogadores, Ganhou, Tentativas, Tempo, Numero).

partida(Jogadores, Ganhou, Tentativas, Tempo, Numero) ->
    io:format("~p~n", [Numero]),
    receive
        tempo -> partida(Jogadores, Ganhou, Tentativas, true, Numero);
        {tentativa, N, Pid} ->
            if
                Tempo ->
                    Pid ! {resposta, "TEMPO\n"},
                    partida(Jogadores, Ganhou, Tentativas, Tempo, Numero);
                Ganhou ->
                    Pid ! {resposta, "PERDEU\n"},
                    partida(Jogadores, Ganhou, Tentativas, Tempo, Numero);
                Tentativas >= 100 ->
                    Pid ! {resposta, "TENTATIVAS\n"},
                    partida(Jogadores, Ganhou, Tentativas, Tempo, Numero);
                true ->
                    if 
                        N == Numero ->
                            Pid ! {resposta, "GANHOU\n"},
                            partida(Jogadores, true, Tentativas + 1, Tempo, Numero);
                        N < Numero ->
                            Pid ! {resposta, "MAIOR\n"},
                            partida(Jogadores, Ganhou, Tentativas + 1, Tempo, Numero);
                        true ->
                            Pid ! {resposta, "MENOR\n"},
                            partida(Jogadores, Ganhou, Tentativas + 1, Tempo, Numero)                            
                    end
            end
    end.

jogador(Server) -> spawn(fun() -> initJogador(Server) end).

initJogador(Server) ->
    Server ! {participa, self()},
    receive
        {start, Partida} -> loopJogador(Partida)
    end.

loopJogador(Partida) ->
    receive
        {jogar, N} ->
            Partida ! {tentativa, N, self()};
        {resposta, Res} -> io:fwrite(Res)
    end,
    loopJogador(Partida).

jogar(Pid, N) -> Pid ! {jogar, N}.